File: src/CPP/src/leetcode/LongestPalindromicSubstring.cpp
/*
 * [Source] https://leetcode.com/problems/longest-palindromic-substring/
 * [Difficulty]: Medium
 * [Tag]: String
 * [Tag]: DP
 */

#include <iostream>

using namespace std;

// [Solution]: Use a boolean matrix which indicates substring between st and ed is palindrom or not. So isPalindrom(st, ed) = str(st) == str(ed) && isPalindrom(st + 1, ed - 1).
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/rcanddp/LongestPalindromicSubstring.java
 */

/* Java solution
public class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() <= 1) {
            return s; 
        }
        boolean[][] f = new boolean[s.length()][s.length()]; 
        int st = 0; 
        int maxLen = 1; 
        int i, j; 
        for (i = 0; i < s.length(); ++i) {
            f[i][i] = true; 
            for (j = 0; j < i; ++j) {
                f[j][i] = s.charAt(j) == s.charAt(i) && (i <= j + 1 || f[j + 1][i - 1]); 
                if (f[j][i] && i - j + 1 >maxLen) {
                    st = j; 
                    maxLen = i - j + 1; 
                }
            }
        }
        return s.substring(st, st + maxLen); 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}

File: wwwroot/js/longest-palindromic-substring.js
/**
 * @param {String} s
 * @return {String}
 */
 const longestPalindrome = function(s) {
    if (s.length <= 1) {
        return s;
    }

    // Use DP.
    let matrix = new Array(s.length + 1);
    for (let i = 0; i < s.length; i++) {
        matrix[i] = new Array(s.length);
    }

    let st = 0;
    let len = 0;
    for (let i = 0; i < s.length; i++) {
        for (let j = 0; j < s.length - i; j++) {
            let val = true;
            if (i === 0) {
                val = true;
            } else if (i === 1) {
                val = (s[j] === s[j + 1] ? true : false);
            }
            else {
                let prev = matrix[j + 1][j + i - 1];
                val = (s[j] === s[j + i] ? prev : false);
            }

            // console.log("j: " + j + " j+i: " + (j + i) + " val: " + val);
            matrix[j][j + i] = val;

            if (val && i + 1 > len) {
                st = j;
                len = i + 1;

                // console.log(s.substr(st, len));
            }
        }
    }

    return s.substr(st, len);
};

document.getElementById("longest-palindromic-substring").innerHTML = longestPalindrome("abacabe");